70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+			};
+			imageBarriers[0].oldLayout = IImage::LAYOUT::UNDEFINED;
+			imageBarriers[0].newLayout = IImage::LAYOUT::GENERAL;
+			imageBarriers[1].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::NONE,
+					.srcAccessMask = ACCESS_FLAGS::NONE,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS
+			}
+			};
+			imageBarriers[1].image = prevVelocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[1].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[1].oldLayout = IImage::LAYOUT::UNDEFINED;
+			imageBarriers[1].newLayout = IImage::LAYOUT::GENERAL;
+
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.imgBarriers = imageBarriers});
+		}
+		else
+		{
+			IGPUCommandBuffer::SPipelineBarrierDependencyInfo::image_barrier_t imageBarriers[2];
+			imageBarriers[0].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS | ACCESS_FLAGS::SHADER_READ_BITS,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS
+			}
+			};
+			imageBarriers[0].image = velocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[0].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[0].oldLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			imageBarriers[0].newLayout = IImage::LAYOUT::GENERAL;
+			imageBarriers[1].barrier = {
+				.dep = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS | ACCESS_FLAGS::SHADER_READ_BITS,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT,
+					.dstAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS
+			}
+			};
+			imageBarriers[1].image = prevVelocityFieldImageView->getCreationParameters().image.get();
+			imageBarriers[1].subresourceRange = {
+				.aspectMask = IImage::EAF_COLOR_BIT,
+				.baseMipLevel = 0u,
+				.levelCount = 1u,
+				.baseArrayLayer = 0u,
+				.layerCount = 1u
+			};
+			imageBarriers[1].oldLayout = IImage::LAYOUT::READ_ONLY_OPTIMAL;
+			imageBarriers[1].newLayout = IImage::LAYOUT::GENERAL;
+
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.imgBarriers = imageBarriers});
+		}
+
+		{
+			SMemoryBarrier memBarrier;
+			memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS;
+			memBarrier.srcAccessMask = ACCESS_FLAGS::MEMORY_WRITE_BITS;
+			memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+			memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+			cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+        }
```

Comment:
you should be barriering against previous COMPUTE and VERTEX stages, why?

Because if you actually did all your init in `onAppInitialized` it would be in a separate submit you await on the CPU with `blockSemaphores` before proceeding to submit anything else, which ensures you have an execution dependency (submit end with semaphore signal performs a massive availability operation while submit begin performs  a massive visibility operation).

Basically as long as you have an execution dep `Sema Signal -> Sema Wait` between submits (even transitively through the host waiting on the sema and holding off on submitting the next commandbuffer till the previous sema signals) its as if you had a massive All_deVICE_STAGES to ALL_DEVICE_STAGES global (all memory) pipeline barrier

Reply:
actually a correct subpass external dependency would take care of last frame vertex barriering against next frame compute.

So here you'd only concern yourself with last compute in last frame barriering against first compute of next frame


Still your dstAccessMask is wrong, should be both READ and WRITE, as you don't want WRITE from previous dispatch accidentally going out of order with a WRITE from next dispatch